control/autoware_control_validator/src/control_validator.cpp

```diff
@@ -56,8 +73,64 @@ bool AccelerationValidator::is_in_error_range() const
          mes >= des - std::abs(e_scale * des) - e_offset;
 }
 
+void VelocityValidator::validate(
+  ControlValidatorStatus & res, const Trajectory & reference_trajectory,
+  const Odometry & kinematics)
+{
+  const double v_vel = vehicle_vel_lpf.filter(kinematics.twist.twist.linear.x);
+  const double t_vel = target_vel_lpf.filter(
+    autoware::motion_utils::calcInterpolatedPoint(reference_trajectory, kinematics.pose.pose)
+      .longitudinal_velocity_mps);
+
+  const bool is_rolling_back =
+    std::signbit(v_vel * t_vel) && std::abs(v_vel) > rolling_back_velocity_th;
+  if (!hold_velocity_error_until_stop || !res.is_rolling_back || std::abs(v_vel) < 0.05) {
+    res.is_rolling_back = is_rolling_back;
+  }
+
+  const bool is_over_velocity =
+    std::abs(v_vel) > std::abs(t_vel) * (1.0 + over_velocity_ratio_th) + over_velocity_offset_th;
+  if (!hold_velocity_error_until_stop || !res.is_over_velocity || std::abs(v_vel) < 0.05) {
+    res.is_over_velocity = is_over_velocity;
+  }
+
+  res.vehicle_vel = v_vel;
+  res.target_vel = t_vel;
+}
+
+void OverrunValidator::validate(
+  ControlValidatorStatus & res, const Trajectory & reference_trajectory,
+  const Odometry & kinematics)
+{
+  res.dist_to_stop = [](const Trajectory & traj, const geometry_msgs::msg::Pose & pose) {
+    const auto stop_idx_opt = autoware::motion_utils::searchZeroVelocityIndex(traj.points);
+
+    const size_t end_idx = stop_idx_opt ? *stop_idx_opt : traj.points.size() - 1;
+    const size_t seg_idx =
+      autoware::motion_utils::findFirstNearestSegmentIndexWithSoftConstraints(traj.points, pose);
+    const double signed_length_on_traj = autoware::motion_utils::calcSignedArcLength(
+      traj.points, pose.position, seg_idx, traj.points.at(end_idx).pose.position,
+      std::min(end_idx, traj.points.size() - 2));
+
+    if (std::isnan(signed_length_on_traj)) {
+      return 0.0;
+    }
+    return signed_length_on_traj;
+  }(reference_trajectory, kinematics.pose.pose);
+
+  res.nearest_trajectory_vel =
+    autoware::motion_utils::calcInterpolatedPoint(reference_trajectory, kinematics.pose.pose)
+      .longitudinal_velocity_mps;
+
+  const double v_vel = vehicle_vel_lpf.filter(kinematics.twist.twist.linear.x);
+
+  // NOTE: the same velocity threshold as autoware::motion_utils::searchZeroVelocity
+  res.has_overrun_stop_point = res.dist_to_stop < -overrun_stop_point_dist_th &&
+                               res.nearest_trajectory_vel < 1e-3 && v_vel > 1e-3;
+}
+
 ControlValidator::ControlValidator(const rclcpp::NodeOptions & options)
-: Node("control_validator", options), validation_params_(), vehicle_info_()
+: Node("control_validator", options), vehicle_info_()
```

Comment:
```suggestion
: Node("control_validator", options), vehicle_info_(),
  latency_validator{*this},
  trajectory_validator{*this},
  // other validator
```

NITS:
By initializing Validators declared as members of the ControlValidator class in the constructor's initialization list, objects are directly constructed, which provides performance benefits.

Reply:
Thank you.